### ISSUE TITLE
Unseting a global variable should report an error.

### ISSUE THREAD

**Issue by 9brada6** [WordPress/WordPress-Coding-Standards#1902]
## Bug Description
Unseting a global variable should report an error.

## Minimal Code Snippet

```php
global $wp_post_types;
$wp_post_types['post'] = $a_variable // Reports an error, nice.
unset( $wp_post_types['post'] ); // Should report an error.
```

## Environment

| Question               | Answer
| ------------------------| -------
| PHP version                     | 7.3
| PHP_CodeSniffer version | 3.5.5
| WPCS version                  | 2.3.0
| WPCS install type            | Composer project local
| IDE (if relevant)                | VSCode

## Additional Context (optional)
I think this is because unsetting a global variable(just the global variable) should not trigger any error, but it wasn't taken into consideration that can be unset a property/key of that global variable.

**Comment by jrfnl** [comment]
Oh, nice. Good point that unsetting variables should be taken into account. I don't think the sniff does, let alone ever has.

> I think this is because unsetting a global variable(just the global variable) should not trigger any error, but it wasn't taken into consideration that can be unset a property/key of that global variable.

Actually, I think the reverse is true.

I.e. unsetting the a WP global variable completely - `unset( $wp_protected_global)` - should trigger an error, while unsetting a key in a WP global variable which is an array - `unset( $wp_protected_global['key'] )` - should not, as there is no telling which keys may have been added by plugins/themes and them unsetting a key they added themselves would be fine.

That is, of course, unless we start keeping track of the expected keys in WP global array variables, but then we'd also need to keep a list of the properties of a WP global object, etc etc., which I don't think is realistic, aside from being a maintenance nightmare.

Does that make sense ?

**Comment by 9brada6** [comment]
Hmm, not so. Here is how I think it should work(for start, and then to improve if have time):

1. For start: Report any unsetting of a global WP variable(including $GLOBALS ofc), including the WP object itself, any array key. object property, ...etc.
2. To make better: Don't report any unsetting of a global reference. Eg: unset( $wp_post_types );

So  step 1, it should be done, step 2 is an improvement that should not be prioritized.

> I.e. unsetting the a WP global variable completely - unset( $wp_protected_global) - should trigger an error

When you declare a variable with global keyword in PHP, that variable is not the actual variable, that is a reference. The actual variable is in $GLOBALS.

```php
global $wp_post_types;   // Make a reference to $GLOBALS['wp_post_types']

unset( $wp_post_types ); // Remove the reference, this does not unset $GLOBALS['wp_post_types'], so variable is unchanged. 
                         // Here it should not **ideally** report an error, since the global variable is not modified, 
                         // but I don't think is a bad idea to report either.
	
global $wp_post_types; // Remake the reference, variable is back again.
```

> while unsetting a key in a WP global variable which is an array - unset( $wp_protected_global['key'] ) - should not, as there is no telling which keys may have been added by plugins/themes and them unsetting a key they added themselves would be fine.

Unsetting a WP global array key(imagine unsetting object property, like $post->ID) is definitely modifying the variable(or overriding as the error suggest), that's just my opinion.

If you make a global variable null, or a string, and an error show up, it should also when unsetting that variable, since unsetting is a biggest modification(if you understand what I try to say).

> as there is no telling which keys may have been added by plugins/themes and them unsetting a key they added themselves would be fine.

The WPCS actually error on that when is added(which is very good), it should when removing. Also adding a property to a WP object or storing variables there is no way a good thing. That is the object of the WP and should be modified by WP not by plugin/themes, plugin/themes should modify with the methods provided, which are plenty.

**Comment by jrfnl** [comment]
@9brada6 You are absolutely right and have reminded me again why I detest (and never use) the `global` keyword.

My response was based on your original code sample, where you were using `global` in the global namespace. In that case, you _are_ altering the original variable.

Though, it is more common for the `global` keyword to be used in a scoped construct, in which case, as you point out, it is a reference and you are modifying the reference, not the original.

Or in code: https://3v4l.org/O5oSF

So any implementation of this, will need to look for:
* `$GLOBALS` with a key pointing to any of the WP global variables being unset *anywhere* and trigger an error.
* A WP global var in the global namespace being unset, completely or partially and trigger an error.
    Mind: this will need to take the `treat_files_as_scoped` property into account (for views being included/required from within a function call).
* A WP global var imported via `global` into a closed scope being unset completely should _not_ trigger an error.
* A WP global var imported via `global` into a closed scope being unset partially (array key/property) should trigger an error.

Do you agree with this summary ?

This would come down to the following:
```php
<?php

function bar() {
    global $wp;

    $wp = ['bar']; // Error: modifying - already in place.
    $wp[] = 'foobar'; // Error: modifying - already in place.
    unset($wp[0]); // Error: unsetting - needs to be added.
    unset($wp); // Ignore.
)

$wp = ['bar']; // Error: modifying - already in place.
$wp[] = 'foobar'; // Error: modifying - already in place.
unset($wp[0]); // Error: unsetting - needs to be added.
unset($wp); // Error: unsetting - needs to be added.

// Behaviour for the code below should depend on the `treat_files_as_scoped` property
// phpcs:set treat_files_as_scoped true
global $wp;

$wp = ['bar']; // Error: modifying - already in place.
$wp[] = 'foobar'; // Error: modifying - already in place.
unset($wp[0]); // Error: unsetting - needs to be added.
unset($wp); // Ignore.

// phpcs:set treat_files_as_scoped false
global $wp;

$wp = ['bar']; // Error: modifying - already in place.
$wp[] = 'foobar'; // Error: modifying - already in place.
unset($wp[0]); // Error: unsetting - needs to be added.
unset($wp); // Error: unsetting - needs to be added.
```

**Comment by 9brada6** [comment]
Yes, it's perfect. Thanks for awesome work for this program.
